src/inference/graph_inference.cpp

```diff
@@ -920,58 +942,90 @@ class GraphNodeAssignment : public Nice2Assignment {
-        if (num_permutations < 0 || num_permutations > FLAGS_permutations_beam_size) {
-          ComputeRandomPermutations(candidate_inf_labels, &permutations, FLAGS_permutations_beam_size, kMaxNumDuplicatesRandomPermutation);
-        } else {
-          ComputeAllPermutations(candidate_inf_labels, &permutations, candidate_inf_labels.size(), FLAGS_permutations_beam_size);
-        }
-        for (auto it = permutations.begin(); it != permutations.end(); ++it) {
-          for (size_t z = 0; z < inf_vars.size(); ++z) {
-            assignments_[inf_vars[z]].label = (*it)[z];
-          }
 
-          bool is_assignment_valid = true;
-          for (size_t z = 0; z < inf_vars.size(); ++z) {
-            if (HasDuplicationConflictsAtNode(inf_vars[z]) ||
-                !fweights.label_checker_.IsLabelValid(assignments_[inf_vars[z]].label)) {
-              is_assignment_valid = false;
-              break;
-            }
-          }
-          if (is_assignment_valid == false) {
-            continue;
+        bool is_assignment_label_valid = true;
+        for (size_t z = 0; z < candidate_inf_labels.size(); ++z) {
+          if (!fweights.label_checker_.IsLabelValid(candidate_inf_labels[z])) {
+            is_assignment_label_valid = false;
           }
-          double score = 0;
-          std::vector<int> assignment;
-          assignment.assign(inf_vars.size(), 0);
-          for (size_t z = 0; z < inf_vars.size(); ++z) {
-            score += GetNodeScore(fweights, inf_vars[z]);
-            assignment[z] = assignments_[inf_vars[z]].label;
+        }
+        if (!is_assignment_label_valid) {
+          continue;
+        }
+        // Perform assignment optimization on the first permutation.
+        PerformPermutationOptimization(inf_nodes, fweights, candidate_inf_labels, &best_assignments, &best_score);
+        uint64 num_permutations = CalculateFactorial(candidate_inf_labels.size());
+        // If the factorial will go in overflow it will return -1.
+        if (num_permutations < 0 || num_permutations > FLAGS_permutations_beam_size) {
+          size_t current_num_permutation = 0;
+          while (current_num_permutation < FLAGS_permutations_beam_size) {
+            std::random_shuffle(candidate_inf_labels.begin(), candidate_inf_labels.end());
+            PerformPermutationOptimization(inf_nodes, fweights, candidate_inf_labels, &best_assignments, &best_score);
+            current_num_permutation++;
           }
-          if (score > best_score) {
-            best_assignments = assignment;
-            best_score = score;
+        } else {
+          std::vector<size_t> count = std::vector<size_t>(candidate_inf_labels.size());
+          size_t current_num_permutations = 0;
+          for (size_t i = 1; i < candidate_inf_labels.size() && current_num_permutations < FLAGS_permutations_beam_size;) {
+            if (count[i] < i) {
+              int swap = i % 2 * count[i];
```

Comment:
Just wondering, now I saw it in details. Are you sure this is a correct way to iterate over the permutations? The standard algorithm is different.

http://marknelson.us/2002/03/01/next-permutation/